src/common/comp/brute_force.cc

```diff
@@ -786,84 +782,88 @@ BruteForce::AnnIterator(const DataSetPtr base_dataset, const DataSetPtr query_da
 #endif
     faiss::MetricType faiss_metric_type = result.value();
     bool is_cosine = IsMetricType(metric_str, metric::COSINE);
-
-    std::unique_ptr<float[]> norms = is_cosine ? GetVecNorms<DataType>(base_dataset) : nullptr;
-    auto pool = ThreadPool::GetGlobalSearchThreadPool();
+    auto larger_is_closer = faiss::is_similarity_metric(faiss_metric_type) || is_cosine;
     auto vec = std::vector<IndexNode::IteratorPtr>(nq, nullptr);
-    std::vector<folly::Future<Status>> futs;
-    futs.reserve(nq);
-
-    for (int i = 0; i < nq; ++i) {
-        futs.emplace_back(pool->push([&, index = i] {
-            ThreadPool::ScopedSearchOmpSetter setter(1);
-
-            BitsetViewIDSelector bw_idselector(bitset, xb_id_offset);
-            [[maybe_unused]] faiss::IDSelector* id_selector = (bitset.empty()) ? nullptr : &bw_idselector;
-            auto larger_is_closer = faiss::is_similarity_metric(faiss_metric_type) || is_cosine;
-            auto max_dis = larger_is_closer ? std::numeric_limits<float>::lowest() : std::numeric_limits<float>::max();
-            std::vector<DistId> distances_ids(nb, {-1, max_dis});
-            [[maybe_unused]] auto cur_query = (const DataType*)xq + dim * index;
-            switch (faiss_metric_type) {
-                case faiss::METRIC_L2: {
-                    if constexpr (std::is_same_v<DataType, knowhere::fp32>) {
-                        faiss::all_L2sqr(cur_query, (const float*)xb, dim, 1, nb, distances_ids, nullptr, id_selector);
-                    } else if constexpr (KnowhereHalfPrecisionFloatPointTypeCheck<DataType>::value) {
-                        faiss::half_precision_floating_point_all_L2sqr(cur_query, (const DataType*)xb, dim, 1, nb,
-                                                                       distances_ids, nullptr, id_selector);
-                    } else {
-                        LOG_KNOWHERE_ERROR_ << "Metric L2 only used in floating point data";
-                        return Status::faiss_inner_error;
-                    }
-                    break;
-                }
-                case faiss::METRIC_INNER_PRODUCT: {
-                    if (is_cosine) {
+    std::shared_ptr<float[]> norms = GetVecNorms<DataType>(base_dataset);
+
+    try {
+        for (int i = 0; i < nq; ++i) {
+            auto compute_dist_func = [=]() -> std::vector<DistId> {
+                auto xb = base_dataset->GetTensor();
```

Comment:
maybe add comments on why we cannot do it outside

Reply:
added.